### PR TITLE
Add SEER preprocessing

### PR DESCRIPTION
**Pull request by Vincent-Maladiere**
Add `hazardous/data/_seer.py` to define the `load_seer` function, which loads the raw SEER txt file and returns the preprocessed X and y.

This file is a helper to reproduce the SEER dataset transformations as performed in SurvTRACE.

### CODE REVIEW

File: hazardous/utils.py [hunk 1]
```diff
@@ -5,10 +5,6 @@
 from sklearn.utils.validation import check_scalar
 
 
-def _dict_to_pd(y):
-    return pd.DataFrame(y)
-
-
```
Comment (ogrisel) [inline hazardous/utils.py:11]: nice simplification :)

File: hazardous/data/_seer.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    7: "ICD-O-3 Hist/behav, malignant",
+    8: "Laterality",
+    9: "Sequence number",
+    10: "Vital status recode (study cutoff used)",
+    11: "ER Status Recode Breast Cancer (1990+)",
+    12: "PR Status Recode Breast Cancer (1990+)",
+    13: "Regional nodes examined (1988+)",
+    14: "xxx",
+    15: "Summary stage 2000 (1998-2017)",
+    16: "xxx",
```
Comment (ogrisel) [inline hazardous/data/_seer.py:49]: Having two different columns with the same `"xxx"` name is likely to cause weird bugs.

Better find descriptive names for those columns, if not possible, then use distinct generic names such as `"Undefined 1"` and `"Undefined 2"`.

File: hazardous/data/_seer.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    19: "First malignant primary indicator",
+    20: "Diagnostic Confirmation",
+    21: "Total number of benign/borderline tumors for patient",
+    22: "Total number of in situ/malignant tumors for patient",
+    23: "Median household income inflation adj to 2019",
+    24: "RX Summ--Surg Prim Site (1998+)",
+    25: "SEER other cause of death classification",
+    26: "SEER cause-specific death classification",
+    27: "Survival months",
+    28: "xxx",
```
Comment (ogrisel) [inline hazardous/data/_seer.py:61]: and another one here.

File: hazardous/data/_seer.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+    Returns
+    -------
+    X : pandas.DataFrame of shape (n_features, n_samples)
+    """
+    X.rename(INDICES_TO_COLUMN_NAMES, axis=1, inplace=True)
+    return X
+
+
+def preprocess_features(X):
```
Comment (ogrisel) [inline hazardous/data/_seer.py:121]: Personally I think this kind of things should not be part of a the dataset loader but instead in the user's code (e.g. in examples or tests), by leveraging the rare categories options of scikit-learn or skrub categorical value encoders. But if we want to keep an option in the dataset loader to ease reproducibility of the paper, let's use a more explicit name such as:

```suggestion
def preprocess_seer_features_as_survtrace(X):
```

File: hazardous/data/_seer.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    X["Median household income inflation adj to 2019"].replace(
+        "Unknown/missing/no match/Not 1990-2018",
+        "$75,000+",
+        inplace=True,
+    )
+
+    X.replace("Unknown", None, inplace=True)
+
+    # TODO: specialize this logic between training and inference.
+    # During inference, we should fetch the trained object.
```
Comment (ogrisel) [inline hazardous/data/_seer.py:178]: This kind of logic should definitely not be part of a dataset loader.

We could instead introduce a `SurvtraceSeerTransformer` with a `fit` and a `transform` method if we really wanted to implement this.

File: hazardous/data/_seer.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    28: "xxx",
+}
+
+
+COLUMN_NAMES_TO_DTYPE = {
+    "Histologic Type ICD-O-3": "int64",
+}
+
+
+def load_seer(input_path, rename=True):
```
Comment (ogrisel) [inline hazardous/data/_seer.py:70]:
```suggestion
def load_seer(input_path, rename=True, survtrace_preprocessing=False):
```

Let's make it possible to enable or disable the survtrace preprocessing. When the survtrace feature preprocessing is disable, this would just get the raw values (with informative column names and meaningful dtypes.

File: hazardous/data/_seer.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+    mask = (
+        (df["SEER cause-specific death classification"] != "N/A not seq 0-59")
+        & (
+            df["Reason no cancer-directed surgery"]
+            != "Not performed, patient died prior to recommended surgery"
+        )
+        & (df["Survival months"] != "Unknown")
+    )
+    df = df.loc[mask]
```
Comment (ogrisel) [inline hazardous/data/_seer.py:96]: I cannot see how this filtering can work when passing `rename=False`. Since the masking code seems to require column names, let's just remove the `rename` kwarg and always do the renaming.

File: hazardous/data/_seer.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+    df_features = preprocess_features(df)
+    df_events = preprocess_events(df)
+
+    df_features = df_features.astype(COLUMN_NAMES_TO_DTYPE)
+
```
Comment (ogrisel) [inline hazardous/data/_seer.py:102]: Please also `.astype` the `CATEGORICAL_COLUMN_NAMES`.

**Review by ogrisel — COMMENTED** [review]
Thanks, here is a first pass of review:

File: hazardous/data/tests/test_seer.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        dict(
+            event=[1.0, 0.0, 1.0],
+            duration=[81, 7, 28],
+        )
+    )
+    assert_frame_equal(y, expected_y)
+    expected_X = pd.read_csv(expected_output_path)
+    valid_columns = X.columns[~X.columns.str.contains("xxx")]
+    assert len(valid_columns) == 25
+    assert_frame_equal(X[valid_columns], expected_X[valid_columns])
```
Comment (ogrisel) [inline hazardous/data/tests/test_seer.py:28]: Please test that `NUMERIC_COLUMN_NAMES` have all numerical dtypes and `CATEGORICAL_COLUMN_NAMES` have categorical dtypes.

Also we could add another test that would use the raw Seer text file and check thinkgs like:

- the number of rows;
- the cardinality of categorical columns and in particular checking that the number of unique values for categorical columns is below some threshold (e.g. 10% of the total number of rows);
- the value counts for each categorical columns;
- the numerical dtypes infers by pandas;
- the mean, median, 25%, 75% percentiles, min and max values of numerical columns;
- that columns that are neither numerical nor categorical have a object dtype (or arrow string in the future but let's not introduce a dependency on pyarrow for now to keep jupyterlite compat in hazardous);
- the number of missing values per columns.

This would help detect that possible behavioral changes in future versions of pandas do not go unnoticed.

File: hazardous/data/_seer.py (hunk 6 above)
Comment (ogrisel) [reply hazardous/data/_seer.py:70]: relative to the hazardous source folder, or in the home dir. If the file does not exist, we could raise a ValueError with a link to the instructions to get the raw text file dataset.

File: hazardous/data/_seer.py (hunk 4 above)
Comment (Vincent-Maladiere) [reply hazardous/data/_seer.py:121]: Makes sense, thanks!

File: hazardous/data/_seer.py (hunk 5 above)
Comment (Vincent-Maladiere) [reply hazardous/data/_seer.py:178]: Whoops, I forgot to remove these. I indeed created SurvTRACE-specific transformers. I'll add these in a subsequent PR.

File: hazardous/data/_seer.py (hunk 6 above)
Comment (Vincent-Maladiere) [reply hazardous/data/_seer.py:70]: So, would you look for a file with a specific name in the home dir, or do you aim at replacing the `FileNotFoundError` with more context? 
WDYT about adding a small guide in the documentation and mentioning it here?

File: hazardous/data/_seer.py (hunk 8 above)
Comment (ogrisel) [reply hazardous/data/_seer.py:102]: Same comment here w.r.t. comment resolution.

File: hazardous/data/_seer.py (hunk 7 above)
Comment (Vincent-Maladiere) [reply hazardous/data/_seer.py:96]: Well, I didn't commit anything new, so nothing has changed. I resolved them in anticipation while working on it, like a checklist.

**Review by ogrisel — APPROVED** [review]
I think this is good enough for a first iteration. We can refine later.

In particular we should retry to download the dataset by following the instructions and document good column names for the unused columns and filter than by explicit name instead of using the "Unused" pattern.

Those are marked as TODO comments in the code.
